indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java

```diff
@@ -1271,6 +1277,113 @@ public void remove(final Task task)
+
+    synchronized (readWriteLock) {
+      try {
+        final boolean acquired = readWriteLock.readLock()
+                                              .tryLock(lockAcquireTimeoutMillis, TimeUnit.MILLISECONDS);
+        if (!acquired) {
+          throw new ISE("Timed out while acquiring transactional append lock for datasource[%s].", datasource);
+        }
+      }
+      catch (InterruptedException e) {
+        throw new ISE(e, "Interrupted while acquiring transactional append lock for datasource[%s].", datasource);
+      }
+    }
+  }
+
+  /**
+   * Acquire a write lock to perform the segment transactional replace action for a given datasource.
+   * Also verifies that all the locks are of the type REPLACE for the task.
+   * @param task task to perform the replace action
+   * @param lockAcquireTimeoutMillis milliseconds to wait for lock acquisition
+   */
+  public void acquireTransactionalReplaceLock(Task task, long lockAcquireTimeoutMillis)
+  {
+    for (TaskLockPosse lockPosse : findLockPossesForTask(task)) {
+      if (lockPosse.taskLock.getType() != TaskLockType.REPLACE) {
+        throw new ISE(
+            "All the locks must be of type REPLACE for segmentTransactionalReplace. Found lock of type[%s] for task[%s].",
+            lockPosse.taskLock.getType(), task.getId()
+        );
+      }
+    }
+
+    final String datasource = task.getDataSource();
+    final ReentrantReadWriteLock readWriteLock =
+        datasourceToConcurrentLock.computeIfAbsent(datasource, ds -> new ReentrantReadWriteLock(true));
+
+    synchronized (readWriteLock) {
+      try {
+        final boolean acquired = readWriteLock.writeLock()
+                                              .tryLock(lockAcquireTimeoutMillis, TimeUnit.MILLISECONDS);
+        if (!acquired) {
+          throw new ISE("Timed out while acquiring transactional replace lock for datasource[%s].", datasource);
+        }
+      }
+      catch (InterruptedException e) {
+        throw new ISE(e, "Interrupted while acquiring transactional replace lock for datasource[%s].", datasource);
+      }
+    }
+  }
+
+  /**
+   * Release the transactional append lock acquired by a task
+   * @param task task to perform the append action
+   */
+  public void releaseTransactionalAppendLock(Task task)
+  {
+    releaseTransactionalLock(task, false);
+  }
+
+  /**
+   * Release the transactional replace lock acquired by a task
+   * @param task task to perform the replace action
+   */
+  public void releaseTransactionalReplaceLock(Task task)
+  {
+    releaseTransactionalLock(task, true);
+  }
+
+  private void releaseTransactionalLock(Task task, boolean replace)
+  {
+    final String datasource = task.getDataSource();
+    final ReentrantReadWriteLock readWriteLock = datasourceToConcurrentLock.get(datasource);
+    synchronized (readWriteLock) {
+      if (!datasourceToConcurrentLock.containsKey(datasource)) {
```

Comment:
Shouldn't this check happen outside the synchronized block? Otherwise, we would synchronize on a `null` object and maybe get an NPE.

Reply:
Thanks, done